File: plugins/lylacore-native/benchmarks/python/benchmark_coach_engine.py
#!/usr/bin/env python3
# Authors: Joysusy & Violet Klaudia 💖
"""
COACH Engine Performance Benchmarks

Comprehensive benchmarking suite for coach-engine Python bindings.
Measures pattern learning, style application, and analysis performance.
"""

import sys
import timeit
from pathlib import Path

# Add parent directory to path to import lylacore
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "crates" / "pyo3-bindings"))

try:
    from lylacore import coach_engine
except ImportError:
    print("Error: lylacore module not found. Build with: maturin develop")
    sys.exit(1)


def benchmark_pattern_learning():
    """Benchmark pattern learning operations."""
    print("\n=== Pattern Learning Benchmarks ===\n")

    user_message = "Hey! How's it going? 😊"
    agent_response = "I'm doing great, thanks for asking!"
    context = {"language": "en", "topic": "greeting"}

    # New pattern
    def bench_new():
        coach_engine.learn_pattern(user_message, agent_response, context)

    iterations = 10000
    time_taken = timeit.timeit(bench_new, number=iterations) / iterations
    print(f"Learn pattern (new)    | {time_taken * 1e6:8.2f} µs")

    # Update existing
    existing_style = coach_engine.learn_pattern(user_message, agent_response, context)

    def bench_update():
        coach_engine.learn_pattern(user_message, agent_response, context, existing_style)

    time_taken = timeit.timeit(bench_update, number=iterations) / iterations
    print(f"Learn pattern (update) | {time_taken * 1e6:8.2f} µs")


def benchmark_style_application():
    """Benchmark style application."""
    print("\n=== Style Application Benchmarks ===\n")

    user_message = "Hey! How's it going? 😊"
    agent_response = "I'm doing great, thanks for asking!"
    context = {"language": "en", "topic": "greeting"}

    style = coach_engine.learn_pattern(user_message, agent_response, context)
    message = "This is a test message for style application"

    def bench():
        coach_engine.apply_style(message, style)

    iterations = 100000
    time_taken = timeit.timeit(bench, number=iterations) / iterations
    print(f"Apply style | {time_taken * 1e9:8.2f} ns")


def benchmark_pattern_merging():
    """Benchmark pattern merging."""
    print("\n=== Pattern Merging Benchmarks ===\n")

    context = {"language": "en", "topic": "general"}

    # Create patterns
    patterns = [
        coach_engine.learn_pattern(f"Message {i}", f"Response {i}", context)
        for i in range(10)
    ]

    for count in [2, 5, 10]:
        def bench():
            coach_engine.merge_patterns(patterns[:count])

        iterations = 10000
        time_taken = timeit.timeit(bench, number=iterations) / iterations
        print(f"Merge {count:2} patterns | {time_taken * 1e6:8.2f} µs")


def benchmark_style_analysis():
    """Benchmark style analysis."""
    print("\n=== Style Analysis Benchmarks ===\n")

    messages = [
        "Hey! How's it going? 😊",
        "What's up?",
        "How are you doing today?",
        "Hope you're having a great day!",
        "Thanks so much for your help!",
    ]

    def bench():
        coach_engine.analyze_style(messages)

    iterations = 10000
    time_taken = timeit.timeit(bench, number=iterations) / iterations
    print(f"Analyze style | {time_taken * 1e6:8.2f} µs")


def main():
    """Run all benchmarks."""
    print("=" * 60)
    print("COACH Engine Python Bindings - Performance Benchmarks")
    print("=" * 60)

    benchmark_pattern_learning()
    benchmark_style_application()
    benchmark_pattern_merging()
    benchmark_style_analysis()

    print("\n" + "=" * 60)
    print("Benchmarks Complete")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: plugins/lylacore-native/benchmarks/nodejs/benchmark_coach_engine.js
// Authors: Joysusy & Violet Klaudia 💖
/**
 * COACH Engine Performance Benchmarks
 *
 * Comprehensive benchmarking suite for coach-engine Node.js bindings.
 * Measures pattern learning, style application, and analysis performance.
 */

const { performance } = require('perf_hooks');

// Import the native addon
let coachEngine;
try {
    coachEngine = require('../../crates/napi-bindings');
} catch (error) {
    console.error('Error: lylacore native addon not found. Build with: npm run build');
    process.exit(1);
}

/**
 * Run a benchmark with adaptive iterations
 */
function benchmark(name, fn, minIterations = 100, targetTime = 1000) {
    // Warm-up
    for (let i = 0; i < 10; i++) {
        fn();
    }

    // Determine iteration count
    const start = performance.now();
    fn();
    const singleTime = performance.now() - start;

    const iterations = Math.max(minIterations, Math.floor(targetTime / singleTime));

    // Actual benchmark
    const benchStart = performance.now();
    for (let i = 0; i < iterations; i++) {
        fn();
    }
    const benchEnd = performance.now();

    const avgTime = (benchEnd - benchStart) / iterations;

    // Format output
    let timeStr;
    if (avgTime < 0.001) {
        timeStr = `${(avgTime * 1e6).toFixed(2)} ns`;
    } else if (avgTime < 1) {
        timeStr = `${(avgTime * 1e3).toFixed(2)} µs`;
    } else {
        timeStr = `${avgTime.toFixed(2)} ms`;
    }

    console.log(`${name.padEnd(30)} | ${timeStr.padStart(12)}`);
}

/**
 * Benchmark pattern learning operations
 */
function benchmarkPatternLearning() {
    console.log('\n=== Pattern Learning Benchmarks ===\n');

    const userMessage = "Hey! How's it going? 😊";
    const agentResponse = "I'm doing great, thanks for asking!";
    const context = { language: 'en', topic: 'greeting' };

    // New pattern
    benchmark('Learn pattern (new)', () => {
        coachEngine.learnPattern(userMessage, agentResponse, context);
    }, 1000);

    // Update existing
    const existingStyle = coachEngine.learnPattern(userMessage, agentResponse, context);

    benchmark('Learn pattern (update)', () => {
        coachEngine.learnPattern(userMessage, agentResponse, context, existingStyle);
    }, 1000);
}

/**
 * Benchmark style application
 */
function benchmarkStyleApplication() {
    console.log('\n=== Style Application Benchmarks ===\n');

    const userMessage = "Hey! How's it going? 😊";
    const agentResponse = "I'm doing great, thanks for asking!";
    const context = { language: 'en', topic: 'greeting' };

    const style = coachEngine.learnPattern(userMessage, agentResponse, context);
    const message = 'This is a test message for style application';

    benchmark('Apply style', () => {
        coachEngine.applyStyle(message, style);
    }, 10000);
}

/**
 * Benchmark pattern merging
 */
function benchmarkPatternMerging() {
    console.log('\n=== Pattern Merging Benchmarks ===\n');

    const context = { language: 'en', topic: 'general' };

    // Create patterns
    const patterns = [];
    for (let i = 0; i < 10; i++) {
        patterns.push(
            coachEngine.learnPattern(`Message ${i}`, `Response ${i}`, context)
        );
    }

    for (const count of [2, 5, 10]) {
        benchmark(`Merge ${count} patterns`, () => {
            coachEngine.mergePatterns(patterns.slice(0, count));
        }, 1000);
    }
}

/**
 * Benchmark style analysis
 */
function benchmarkStyleAnalysis() {
    console.log('\n=== Style Analysis Benchmarks ===\n');

    const messages = [
        "Hey! How's it going? 😊",
        "What's up?",
        "How are you doing today?",
        "Hope you're having a great day!",
        "Thanks so much for your help!",
    ];

    benchmark('Analyze style', () => {
        coachEngine.analyzeStyle(messages);
    }, 1000);
}

/**
 * Main benchmark runner
 */
function main() {
    console.log('='.repeat(60));
    console.log('COACH Engine Node.js Bindings - Performance Benchmarks');
    console.log('='.repeat(60));

    benchmarkPatternLearning();
    benchmarkStyleApplication();
    benchmarkPatternMerging();
    benchmarkStyleAnalysis();

    console.log('\n' + '='.repeat(60));
    console.log('Benchmarks Complete');
    console.log('='.repeat(60));
}

// Run benchmarks
main();
